### PR TITLE
server: allow listing custom offerings for a running VM

### LINKED ISSUE #5412: Cannot select constrained or unconstrained offering on powered on VM
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
UI
~~~

##### CLOUDSTACK VERSION

~~~
4.15.1
~~~

##### OS / ENVIRONMENT

VMware

##### STEPS TO REPRODUCE

~~~
In new and legacy UI Power on VM in Vmware environment and try to scale it .
CPU and Memory hot add is enabled.
~~~
Screenshot 2021-09-07 134933

Screenshot 2021-09-07 134814

##### EXPECTED RESULTS

~~~
Scaling can be done with constrained and unconstrained offerings.
~~~

##### ACTUAL RESULTS

~~~
Scaling cant be done with constrained and unconstrained offerings.
~~~

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

`listServiceOffering` API when called for a running VM only fails to list all possible offerings especially the custom offerings and all offerings when VM is deployed using a custom offering.
This PR improves the list behviour by adding adding appropriate conditions for the custom offerings.

Fixes #5412

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

Offerings and VMs
```
(local) 🦐 > list serviceofferings filter=id,name,cpunumber,cpuspeed,memory,serviceofferingdetails
{
  "count": 6,
  "serviceoffering": [
    {
      "cpunumber": 1,
      "cpuspeed": 500,
      "id": "c17c21bb-faa4-4c14-9774-142260546670",
      "memory": 512,
      "name": "Small Instance"
    },
    {
      "cpunumber": 1,
      "cpuspeed": 1000,
      "id": "0a785dc5-ff8d-48ed-82b2-e24cbd9587b9",
      "memory": 1024,
      "name": "Medium Instance"
    },
    {
      "cpunumber": 1,
      "cpuspeed": 1000,
      "id": "3801bb99-4077-4685-919f-2ae51fce90a9",
      "memory": 1536,
      "name": "fixed"
    },
    {
      "cpuspeed": 1000,
      "id": "fe7de1af-b045-4749-a88f-c20a50082c7b",
      "name": "Constrained",
      "serviceofferingdetails": {
        "maxcpunumber": "4",
        "maxmemory": "2560",
        "mincpunumber": "1",
        "minmemory": "256"
      }
    },
    {
      "id": "3799cd94-8284-485e-81f0-bc3c1e6a29f9",
      "name": "Unconstrained"
    },
    {
      "cpuspeed": 1000,
      "id": "b1423e9e-9fcf-48c1-b156-b563cf9bf4dd",
      "name": "Constrained-1",
      "serviceofferingdetails": {
        "maxcpunumber": "4",
        "maxmemory": "3072",
        "mincpunumber": "2",
        "minmemory": "512"
      }
    }
  ]
}
(local) 🦐 > list virtualmachines ids=112b9b4a-12e9-4cfc-8bd7-0e49eabb211f,00774881-64cf-45d1-8c9c-7bdaf7f195da filter=id,name,serviceofferingid,serviceofferingname,details,state
{
  "count": 2,
  "virtualmachine": [
    {
      "details": {
        "cpuNumber": "2",
        "memory": "1024"
      },
      "id": "00774881-64cf-45d1-8c9c-7bdaf7f195da",
      "name": "t5",
      "serviceofferingid": "fe7de1af-b045-4749-a88f-c20a50082c7b",
      "serviceofferingname": "Constrained",
      "state": "Running"
    },
    {
      "details": {},
      "id": "112b9b4a-12e9-4cfc-8bd7-0e49eabb211f",
      "name": "t3",
      "serviceofferingid": "c17c21bb-faa4-4c14-9774-142260546670",
      "serviceofferingname": "Small Instance",
      "state": "Running"
    }
  ]
}
````

Before change:
```
(local) 🐣 > list serviceofferings filter=id,name,cpunumber,cpuspeed,memory,serviceofferingdetails virtualmachineid=112b9b4a-12e9-4cfc-8bd7-0e49eabb211f 
{
  "count": 2,
  "serviceoffering": [
    {
      "cpunumber": 1,
      "cpuspeed": 1000,
      "id": "0a785dc5-ff8d-48ed-82b2-e24cbd9587b9",
      "memory": 1024,
      "name": "Medium Instance"
    },
    {
      "cpunumber": 1,
      "cpuspeed": 1000,
      "id": "3801bb99-4077-4685-919f-2ae51fce90a9",
      "memory": 1536,
      "name": "fixed"
    }
  ]
}
(local) 🐓 > list serviceofferings filter=id,name,cpunumber,cpuspeed,memory,serviceofferingdetails virtualmachineid=00774881-64cf-45d1-8c9c-7bdaf7f195da 
```

After change:
```
(local) 🐕 > list serviceofferings filter=id,name,cpunumber,cpuspeed,memory,serviceofferingdetails virtualmachineid=112b9b4a-12e9-4cfc-8bd7-0e49eabb211f 
{
  "count": 5,
  "serviceoffering": [
    {
      "cpunumber": 1,
      "cpuspeed": 1000,
      "id": "0a785dc5-ff8d-48ed-82b2-e24cbd9587b9",
      "memory": 1024,
      "name": "Medium Instance"
    },
    {
      "cpunumber": 1,
      "cpuspeed": 1000,
      "id": "3801bb99-4077-4685-919f-2ae51fce90a9",
      "memory": 1536,
      "name": "fixed"
    },
    {
      "cpuspeed": 1000,
      "id": "fe7de1af-b045-4749-a88f-c20a50082c7b",
      "name": "Constrained",
      "serviceofferingdetails": {
        "maxcpunumber": "4",
        "maxmemory": "2560",
        "mincpunumber": "1",
        "minmemory": "256"
      }
    },
    {
      "id": "3799cd94-8284-485e-81f0-bc3c1e6a29f9",
      "name": "Unconstrained"
    },
    {
      "cpuspeed": 1000,
      "id": "b1423e9e-9fcf-48c1-b156-b563cf9bf4dd",
      "name": "Constrained-1",
      "serviceofferingdetails": {
        "maxcpunumber": "4",
        "maxmemory": "3072",
        "mincpunumber": "2",
        "minmemory": "512"
      }
    }
  ]
}
(local) 🦑 > list serviceofferings filter=id,name,cpunumber,cpuspeed,memory,serviceofferingdetails virtualmachineid=00774881-64cf-45d1-8c9c-7bdaf7f195da 
{
  "count": 3,
  "serviceoffering": [
    {
      "cpuspeed": 1000,
      "id": "fe7de1af-b045-4749-a88f-c20a50082c7b",
      "name": "Constrained",
      "serviceofferingdetails": {
        "maxcpunumber": "4",
        "maxmemory": "2560",
        "mincpunumber": "1",
        "minmemory": "256"
      }
    },
    {
      "id": "3799cd94-8284-485e-81f0-bc3c1e6a29f9",
      "name": "Unconstrained"
    },
    {
      "cpuspeed": 1000,
      "id": "b1423e9e-9fcf-48c1-b156-b563cf9bf4dd",
      "name": "Constrained-1",
      "serviceofferingdetails": {
        "maxcpunumber": "4",
        "maxmemory": "3072",
        "mincpunumber": "2",
        "minmemory": "512"
      }
    }
  ]
}
```

### CODE REVIEW

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@blueorangutan test

**Review by rhtyd — APPROVED** [review]
LGTM - did not test the PR

**Review by nvazquez — APPROVED** [review]
Code LGTM

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by rhtyd** [comment]
ping @shwstppr - can you fix the conflict?

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/api/query/QueryManagerImpl.java [hunk 1]
```diff
@@ -2994,9 +3034,28 @@
 
             // 2.In case vm is running return only offerings greater than equal to current offering compute and offering's dynamic scalability should match
             if (vmInstance.getState() == VirtualMachine.State.Running) {
-                sc.addAnd("cpu", Op.GTEQ, currentVmOffering.getCpu());
-                sc.addAnd("speed", Op.GTEQ, currentVmOffering.getSpeed());
-                sc.addAnd("ramSize", Op.GTEQ, currentVmOffering.getRamSize());
+                Integer vmCpu = currentVmOffering.getCpu();
+                Integer vmMemory = currentVmOffering.getSpeed();
```
Comment (davidjumani) [inline server/src/main/java/com/cloud/api/query/QueryManagerImpl.java:3038]:
```suggestion
                Integer vmMemory = currentVmOffering.getRamSize();
```

File: server/src/main/java/com/cloud/api/query/QueryManagerImpl.java [hunk 2]
```diff
@@ -2994,9 +3034,28 @@
...
+                    UserVmVO userVmVO = _userVmDao.findById(vmId);
+                    _userVmDao.loadDetails(userVmVO);
+                    Map<String, String> details = userVmVO.getDetails();
+                    vmCpu = NumbersUtil.parseInt(details.get(ApiConstants.CPU_NUMBER), 0);
+                    if (vmSpeed == null) {
+                        vmSpeed = NumbersUtil.parseInt(details.get(ApiConstants.CPU_SPEED), 0);
+                    }
+                    vmMemory = NumbersUtil.parseInt(details.get(ApiConstants.MEMORY), 0);
+                }
+                if (vmCpu != null && vmCpu > 0) {
```
Comment (davidjumani) [inline server/src/main/java/com/cloud/api/query/QueryManagerImpl.java:3050]: Would it be better to have a single method to take cpu number, speed and memory and return a single sc instead ?

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-67u3

File: server/src/main/java/com/cloud/api/query/QueryManagerImpl.java (hunk 2 above)
Comment (shwstppr) [reply server/src/main/java/com/cloud/api/query/QueryManagerImpl.java:3050]: Not sure if that would add any benefit as that single SearchCriteria will also be a combination of these three SearchCriterias with AND condition

**Comment by shwstppr** [comment]
@rhtyd will restart tests as had to make a change
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test centos7 vmware-67u3

**Review by sureshanaparti — APPROVED** [review]

**Comment by sureshanaparti** [comment]
Merging based on approvals and test results

**Pull request merged**